astropy/table/table.py

```diff
@@ -887,12 +887,7 @@ def _init_from_list_of_dicts(self, data, names, dtype, n_cols, copy):
         for row in data:
             names_from_data.update(row)
 
-        # Put names into a preferred order, either using the first row of data
-        # if it is ordered, or alphabetically.  Starting with Python 3.7, dict
-        # is ordered so this test can be relaxed.  (In practice CPython 3.6 is
-        # this way, but not according to the formal spec).
-        if (isinstance(data[0], OrderedDict)
-                and set(data[0].keys()) == names_from_data):
+        if set(data[0].keys()) == names_from_data:
```

Comment:
Not sure if I completely understand the comment, so @taldcroft and @mhvk might want to double check.

Reply:
Looks good!

Reply:
Looks like this caused real failures. What does this mean?
```
____________________ test_init_from_row_OrderedDict[False] _____________________
table_type = <class 'astropy.table.table.Table'>
    @pytest.mark.usefixtures('table_type')
    def test_init_from_row_OrderedDict(table_type):
        row1 = OrderedDict([('b', 1), ('a', 0)])
        row2 = {'a': 10, 'b': 20}
        rows12 = [row1, row2]
        row3 = dict([('b', 1), ('a', 0)])
        row4 = dict([('b', 11), ('a', 10)])
        rows34 = [row3, row4]
        t1 = table_type(rows=rows12)
        t2 = table_type(rows=rows34)
        assert t1.colnames == ['b', 'a']
>       assert t2.colnames == ['a', 'b']
E       AssertionError: assert ['b', 'a'] == ['a', 'b']
E         At index 0 diff: 'b' != 'a'
E         Use -v to get the full diff
../../.tox/py37-test/lib/python3.7/site-packages/astropy/table/tests/test_init_table.py:550: AssertionError
_____________________ test_init_from_row_OrderedDict[True] _____________________
table_type = <class 'astropy.table.tests.conftest.MaskedTable'>
    @pytest.mark.usefixtures('table_type')
    def test_init_from_row_OrderedDict(table_type):
        row1 = OrderedDict([('b', 1), ('a', 0)])
        row2 = {'a': 10, 'b': 20}
        rows12 = [row1, row2]
        row3 = dict([('b', 1), ('a', 0)])
        row4 = dict([('b', 11), ('a', 10)])
        rows34 = [row3, row4]
        t1 = table_type(rows=rows12)
        t2 = table_type(rows=rows34)
        assert t1.colnames == ['b', 'a']
>       assert t2.colnames == ['a', 'b']
E       AssertionError: assert ['b', 'a'] == ['a', 'b']
E         At index 0 diff: 'b' != 'a'
E         Use -v to get the full diff
../../.tox/py37-test/lib/python3.7/site-packages/astropy/table/tests/test_init_table.py:550: AssertionError
```

Reply:
So it means that the non-ordered dict didn't keep the column names in the right order. Odd. Need to look at it in a bit more detail...

Reply:
`dict` preserves insertion order now, so the new answer `['b', 'a']` looks legit. I am just not sure about the intention of your logic.

Reply:
Ah, the problem was that before the columns were sorted by name since it was not an `OrderedDict`, but now we can do better. So, it seems like we should adjust the test. Pinging @taldcroft since this may change things for users.

Reply:
Agreed we should just adjust the test, but there needs to be an API change entry indicating the update.
```
The order of columns when creating a table from a ``list`` of ``dict`` may be changed.
Previously the order was alphabetical because the ``dict`` keys were assumed to be
in random order. Since Python 3.7 the keys are always in order of insertion so ``Table``
now uses the order of keys in the first row to set the column order. To alphabetize the
columns to match the previous behavior, use ``t = t[sorted(t.colnames)]``.
```

Reply:
Thanks! I pushed a commit for this.